grpctest/grpctest.go

```diff
@@ -0,0 +1,80 @@
+/*
+Package grpctest provides utilities for RPC testing.
+
+Call grpctest.NewServer to create a Server for a test;
+use its Addr field as the address to dial, or invoke its Dial method.
+*/
+package grpctest
+
+import (
+	"net"
+	"testing"
+
+	"google.golang.org/grpc"
+)
+
+// A Server is an RPC server listening on a system-chosen port on the local loopback interface.
+type Server struct {
+	GRPC *grpc.Server
+	Addr string // address of test server, suitable for passing to grpc.Dial
+
+	t       *testing.T
+	l       net.Listener
+	closing bool
+	closed  chan struct{}
+}
+
+// NewServer returns a new unstarted Server.
+// Attach RPC services to s.GRPC, then invoke s.Start to start the server.
+// The caller should call the Server's Close method when finished.
+func NewServer(t *testing.T) *Server {
```

Comment:
This uses and encourages an anti-pattern for Go testing, which is passing `testing.T`s around (which is particularly discouraged when crossing package boundaries and storing in structs AIUI).

Further, I'm not convinced this package is worth the extra API surface it introduces.  It is easy to start and stop a server/client with grpc.  The main complexity is the `Listen`-on-localhost problem (which really should be solved in the `net` package itself, but that's irrelevant here).  In the not-too-distant future (~1Q or so), we should be introducing an in-process transport that will be even easier to serve/dial.

Reply:
It's fine to pass around `testing.T` in limited circumstances. That's why `t.Helper` exists, after all. It solves the problems that passing around a `testing.T` produces.

It may be easy to start and stop a server for people like us who know what's going on, but lots of people aren't so comfortable. I see a lot of people getting it wrong in various ways (e.g. binding to `:0` instead of a loopback address, or starting the server before registering service handlers). It's not a lot of code, but it cuts down the error rate a lot.